### PR TITLE
P2441R2 views::join_with

### LINKED ISSUE #5270: [2022-02 LWG Motion 11] P2441R2 views::join_with
[P2441R2](https://wg21.link/p2441r2)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #5270
Fixes cplusplus/papers#1107

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -5804,6 +5816,581 @@
...
+  requires @\exposid{ref-is-glvalue}@ && @\libconcept{bidirectional_range}@<@\exposid{Base}@> &&
+           @\exposconcept{bidirectional-common}@<@\exposid{InnerBase}@> && @\exposconcept{bidirectional-common}@<@\exposid{PatternBase}@>;
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+iterator tmp = *this;
```
Comment (JohelEGP) [inline source/ranges.tex:6315]:
```suggestion
@\exposid{iterator}@ tmp = *this;
```

File: source/ranges.tex [hunk 2]
```diff
@@ -5804,6 +5816,581 @@
...
+      requires Const && @\libconcept{convertible_to}@<sentinel_t<V>, sentinel_t<@\exposid{Base}@>>;
+
+    template <bool OtherConst>
+      requires @\libconcept{sentinel_for}@<sentinel_t<@\exposid{Base}@>, iterator_t<@\exposid{maybe-const}@<OtherConst, V>>>
+    friend constexpr bool operator==(const @\exposid{iterator}@<OtherConst>& x, const @\exposid{sentinel}@& y);
+};
+\end{codeblock}
+
+\begin{itemdecl}
+constexpr explicit sentinel(Parent& parent);
```
Comment (JohelEGP) [inline source/ranges.tex:6362]:
```suggestion
constexpr explicit @\exposid{sentinel}@(@\exposid{Parent}@& parent);
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/ranges.tex [hunk 3]
```diff
@@ -5804,6 +5816,581 @@
...
+  class join_with_view : public view_interface<join_with_view<V, Pattern>> {
+    using @\exposid{InnerRng}@ = range_reference_t<V>;                  // \expos
+
+    V @\exposid{base_}@ = V();                                          // \expos
+    @\exposid{non-propagating-cache}@<remove_cv_t<@\exposid{InnerRng}@>> @\exposid{inner_}@;   // \expos, present only
+                                                            // when \tcode{!is_reference_v<\exposid{InnerRng}>}
+    Pattern @\exposid{pattern_}@ = Pattern();                           // \expos
+
+    template<bool Const> struct @\exposid{iterator}@;                   // \expos
+    template<bool Const> struct @\exposid{sentinel}@;                   // \expos
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/ranges.tex]: fixed

File: source/ranges.tex [hunk 4]
```diff
@@ -5804,6 +5816,581 @@
...
+                    is_reference_v<InnerRng> && @\libconcept{forward_range}@<@\exposid{InnerRng}@> &&
+                    @\libconcept{common_range}@<V> && @\libconcept{common_range}@<@\exposid{InnerRng}@>)
+        return @\exposid{iterator}@<@\exposconcept{simple-view}@<V> && @\exposconcept{simple-view}@<Pattern>>{*this, ranges::end(@\exposid{base_}@)};
+      else
+        return @\exposid{sentinel}@<@\exposconcept{simple-view}@<V> && @\exposconcept{simple-view}@<Pattern>>{*this};
+    }
+    constexpr auto end() const
+      requires @\libconcept{input_range}@<const V> && @\libconcept{forward_range}@<const Pattern> &&
+               is_reference_v<range_reference_t<const V>> {
+      using @\exposid{InnerConstRng}@ = range_reference_t<const V>;
```
[the remark this answers is not in the document]
Comment (jensmaurer) [reply source/ranges.tex]: removed `\exposid`

File: source/ranges.tex [hunk 5]
```diff
@@ -5804,6 +5816,583 @@
...
+    using @\exposid{InnerRng}@ = range_reference_t<V>;                  // \expos
+
+    V @\exposid{base_}@ = V();                                          // \expos
+    @\exposid{non-propagating-cache}@<remove_cv_t<@\exposid{InnerRng}@>> @\exposid{inner_}@;   // \expos, present only
+                                                            // when \tcode{!is_reference_v<\exposid{InnerRng}>}
+    Pattern @\exposid{pattern_}@ = Pattern();                           // \expos
+
+    // \ref{range.join.with.iterator}, class \tcode{join_with_view::\exposid{iterator}}
+    template<bool Const> struct @\exposid{iterator}@;                   // \expos
+    // \ref{range.join.with.sentinel}, class \tcode{join_with_view::\exposid{sentinel}}
```
Comment (JohelEGP) [inline source/ranges.tex:5876]:
```suggestion
    // \ref{range.join.with.iterator}, class template \tcode{join_with_view::\exposid{iterator}}
    template<bool Const> struct @\exposid{iterator}@;                   // \expos
    // \ref{range.join.with.sentinel}, class template \tcode{join_with_view::\exposid{sentinel}}
```

File: source/ranges.tex [hunk 6]
```diff
@@ -5804,6 +5816,583 @@
...
+        return @\exposid{iterator}@<@\exposconcept{simple-view}@<V> && @\exposconcept{simple-view}@<Pattern>>{*this, ranges::end(@\exposid{base_}@)};
+      else
+        return @\exposid{sentinel}@<@\exposconcept{simple-view}@<V> && @\exposconcept{simple-view}@<Pattern>>{*this};
+    }
+    constexpr auto end() const
+      requires @\libconcept{input_range}@<const V> && @\libconcept{forward_range}@<const Pattern> &&
+               is_reference_v<range_reference_t<const V>> {
+      using InnerConstRng = range_reference_t<const V>;
+      if constexpr (@\libconcept{forward_range}@<const V> && @\libconcept{forward_range}@<@\exposid{InnerConstRng}@> &&
+                    @\libconcept{common_range}@<const V> && common_range<@\exposid{InnerConstRng}@>)
```
Comment (JohelEGP) [inline source/ranges.tex:5917]:
```suggestion
      if constexpr (@\libconcept{forward_range}@<const V> && @\libconcept{forward_range}@<@InnerConstRng> &&
                    @\libconcept{common_range}@<const V> && common_range<InnerConstRng>)
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

**Pull request merged**